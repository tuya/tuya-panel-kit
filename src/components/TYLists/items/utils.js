/* eslint-disable no-param-reassign */

/**
 * https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_pick
 */
export const pick = (object, keys) => {
  return keys.reduce((obj, key) => {
    if (typeof object[key] !== 'undefined') {
      obj[key] = object[key];
    }
    return obj;
  }, {});
};

export const omit = (object, keys) => {
  const shallowCopy = {
    ...object,
  };
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    delete shallowCopy[key];
  }
  return shallowCopy;
};
