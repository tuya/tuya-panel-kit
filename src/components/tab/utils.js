import React from 'react';

const toArray = children => {
  const arr = [];
  React.Children.forEach(children, child => {
    if (child) {
      arr.push(child);
    }
  });
  return arr;
};

const getActiveIndex = (children, activeKey) => {
  const arr = toArray(children);
  for (let i = 0; i < arr.length; i++) {
    if (arr[i].key === activeKey) {
      return i;
    }
  }
  return -1;
};

const activeKeyIsValid = (children, key) => {
  const keys = React.Children.map(children, child => child && child.key);
  return keys.indexOf(key) >= 0;
};

const getDefaultActiveKey = children => {
  let activeKey;
  React.Children.forEach(children, child => {
    if (child && !activeKey && !child.props.disabled) {
      activeKey = child.key;
    }
  });
  return activeKey;
};

export default {
  toArray,
  getActiveIndex,
  activeKeyIsValid,
  getDefaultActiveKey,
};
