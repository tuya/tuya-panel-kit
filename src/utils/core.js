/* eslint-disable no-param-reassign */
/**
 * @example
 * toFixed('111', 5)
 * // '00111'
 * toFixed('3456111', 5)
 * // '56111'
 * @param {String/Number} str is a string
 * @param {Number} count is number, the length of the return string
 * @returns {String}
 */
const toFixed = (str, count) => {
  return `${'0'.repeat(count)}${str}`.slice(-1 * count);
};

/**
 * @example
 * toFilled('111', 5)
 * // '00111'
 * toFilled('3456111', 5)
 * // '3456111'
 * @param {String} str is a string
 * @param {Number} count is number, the length of the return string
 * @returns {String}
 */
const toFilled = (str, count) => {
  let s = str;
  if (s.length < count) {
    s = '0'.repeat(count - s.length) + s;
  }
  return s;
};

/**
 * @example
 * partition('1234567', 3)
 * // ['123', '456', '7']
 * @param {String} str is a string
 * @param {Number} chunk is number, the length of the return string
 * @returns {Array} a array, item of it is a substring which length is the `chunk`
 */
const partition = (str, chunk) => {
  const res = [];
  const len = str.length;
  for (let i = 0; i < len; i += chunk) {
    res.push(str.slice(i, i + chunk));
  }
  return res;
};

/**
 * @example
 * isObject('1234567', 3)
 * // ['123', '456', '7']
 * @param {Any} obj is a anything
 * @returns {Boolean}
 */
const isObject = obj => obj === Object(obj);

const isArray = obj => Object.prototype.toString.call(obj) === '[object Array]';

const isDate = obj => Object.prototype.toString.call(obj) === '[object Date]';

const isRegExp = obj => Object.prototype.toString.call(obj) === '[object RegExp]';

const isBoolean = obj => Object.prototype.toString.call(obj) === '[object Boolean]';

const isNumerical = obj => Object.prototype.toString.call(obj) === '[object Number]';

const isUndefined = obj => typeof obj === 'undefined';

const isNil = obj => obj === undefined || obj === null;

/**
 * lodash get polyfill
 * https://gist.github.com/dfkaye/59263b51cf1e0b633181c5f44ae2066a
 */
const get = function(object, pathString, defaultValue) {
  // Coerce pathString to a string (even it turns into "[object Object]").
  const parts = `${pathString}`.split('.');
  const { length } = parts;
  let i = 0;

  // In case object isn't a real object, set it to undefined.
  let value = object === Object(object) ? object : undefined;

  while (value != null && i < length) {
    value = value[parts[i++]];
  }

  /**
   * lodash.get() returns the resolved value if
   * 1. iteration happened (i > 0)
   * 2. iteration completed (i === length)
   * 3. the value at the path is found in the data structure (not undefined). Note that if the path is found but the
   *    value is null, then null is returned.
   * If any of those checks fails, return the defaultValue param, if provided.
   */
  return i && i === length && value !== undefined ? value : defaultValue;
};

/**
 * https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_pick
 */
const pick = (object, keys) => {
  return keys.reduce((obj, key) => {
    if (typeof object[key] !== 'undefined') {
      obj[key] = object[key];
    }
    return obj;
  }, {});
};

const omit = (object, keys) => {
  const shallowCopy = {
    ...object,
  };
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    delete shallowCopy[key];
  }
  return shallowCopy;
};

const chunk = (arr, chunkSize = 1, cache = []) => {
  const tmp = [...arr];
  if (chunkSize <= 0) return cache;
  while (tmp.length) cache.push(tmp.splice(0, chunkSize));
  return cache;
};

const compareVersion = (v1, v2) => {
  if (typeof v1 !== 'string') return false;
  if (typeof v2 !== 'string') return false;
  v1 = v1.split('.');
  v2 = v2.split('.');
  const k = Math.min(v1.length, v2.length);
  for (let i = 0; i < k; ++i) {
    v1[i] = parseInt(v1[i], 10);
    v2[i] = parseInt(v2[i], 10);
    if (v1[i] > v2[i]) return 1;
    if (v1[i] < v2[i]) return -1;
  }
  return v1.length === v2.length ? 0 : v1.length < v2.length ? -1 : 1;
};

export default {
  get,
  toFixed,
  toFilled,
  partition,
  isObject,
  isArray,
  isDate,
  isRegExp,
  isBoolean,
  isNumerical,
  isUndefined,
  isNil,
  pick,
  omit,
  chunk,
  compareVersion,
};
