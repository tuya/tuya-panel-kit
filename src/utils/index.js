import ColorUtils from './color';
import CoreUtils from './core';
import JsonUtils from './json';
import NumberUtils from './number';
import RatioUtils from './ratio';
import StringUtils from './string';
import TemperatureUtils from './temperature';
import TimeUtils from './time';
import ThemeUtils from './theme';

export {
  ColorUtils,
  CoreUtils,
  JsonUtils,
  NumberUtils,
  RatioUtils,
  StringUtils,
  TemperatureUtils,
  TimeUtils,
  ThemeUtils,
};
