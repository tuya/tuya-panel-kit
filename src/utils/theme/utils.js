import { css } from 'styled-components/native';
import CoreUtils from '../core';

const { get, isNil } = CoreUtils;

export const isObject = item => item && typeof item === 'object' && !Array.isArray(item);

export const deepMerge = (target, ...sources) => {
  if (!sources.length) {
    return target;
  }
  // making sure to not change target (immutable)
  const output = { ...target };
  sources.forEach(source => {
    if (isObject(source)) {
      Object.keys(source).forEach(key => {
        if (isObject(source[key]) && isObject(output[key])) {
          if (!output[key]) {
            output[key] = { ...source[key] };
          } else {
            output[key] = deepMerge(output[key], source[key]);
          }
        } else {
          output[key] = source[key];
        }
      });
    }
  });
  return output;
};

/**
 * @desc 解析上下左右值
 *
 * @param {Array} values - 上右下左值
 * @param {String} key - 样式key，如margin、padding等
 */
export const parseToStyle = (values, key) => {
  if (!values || !Array.isArray(values)) {
    return {};
  }
  const [top, right, bottom, left] = values;
  return {
    [`${key}Top`]: top,
    [`${key}Right`]: right,
    [`${key}Bottom`]: bottom,
    [`${key}Left`]: left,
  };
};

/**
 * @desc 解析上下左右值
 *
 * @param {Array} values - 上右下左值
 * @param {String} key - 样式key，如margin、padding等
 */
export const parseToCss = (values, key) => {
  if (!values || !Array.isArray(values)) {
    return [];
  }
  const [top, right, bottom, left] = values;
  const cssStyle = css`
    ${key}-top: ${top}px;
    ${key}-right: ${right}px;
    ${key}-bottom: ${bottom}px;
    ${key}-left: ${left}px;
  `;
  return cssStyle;
};

/**
 * @desc 从 theme 中取需要的 `theme` 值，
 *
 * @param {Object} props - props，必须包含theme
 * @param {String} key - theme key
 * @param {String | Number} defaultValue - 默认主题值
 *
 * @example
 *
 * 优先级: (以Slider为例)
 * getTheme(props, 'slider.thumbTintColor', '#fff');
 * 
 * props.thumbTintColor ->
 * props.theme.slider.thumbTintColor ->
 * props.theme.slider[props.theme.slider.type].thumbTintColor ->
 * props.theme.slider[props.theme.type].thumbTintColor

 *
 * ```javascript
 *  const StyledView = styled(View)`
 *    background-color: ${props => getTheme(props, 'slider.thumbTintColor', defaultTheme.slider.light.thumbTintColor)};
 * `;
 * ```
 *
 * ```javascript
 *  const defaultTheme = {
 *    type: 'light',
 *    slider: {
 *       type: 'dark',
 *       light: {
 *         thumbTintColor: '#fff',
 *       },
 *       dark: {
 *         thumbTintColor: '#333',
 *         // thumbTintColor: props => props.disabled ? '#999' : '#fff',
 *       }
 *    }
 *  }
 * ```
 */
export const getTheme = (props, key, defaultValue) => {
  const themePath = typeof key === 'function' ? key(props) : key;
  const [namespace, ...path] = key.split('.');
  const themeBasicPath = path.join('.');
  const themeType = get(props.theme, `${namespace}.type`) || get(props.theme, 'type', 'light');
  let themeValue, themeTypePath;
  if (themeType) {
    themeTypePath = [namespace, themeType, ...path].join('.');
  }
  if (!isNil(get(props, themeBasicPath))) {
    themeValue = get(props, themeBasicPath);
  } else if (!isNil(get(props.theme, themePath))) {
    themeValue = get(props.theme, themePath);
  } else {
    themeValue = get(props.theme, themeTypePath, defaultValue);
  }
  if (typeof themeValue === 'function') {
    themeValue = themeValue(props);
  }
  return themeValue;
};

export const createGetTheme = namespace => (key, defaultValue) => {
  const themeKey = `${namespace}.${key}`;
  return getTheme(themeKey, defaultValue);
};
