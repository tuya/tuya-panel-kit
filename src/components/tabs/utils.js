import { RatioUtils, NumberUtils } from '../../utils';
import { PRESS_THRESHOLD, SWIPE_THRESHOLD, VELOCITY_THRESHOLD } from './constant';

const { winWidth } = RatioUtils;
const { inMaxMin } = NumberUtils;

/**
 * @desc 计算tab的宽度
 *
 * @param {Number} maxNum - 一屏上的tab最大数量
 */
export const getTabWidth = (maxNum = 4, tabsWidth = winWidth) => tabsWidth / maxNum;

/**
 * @desc 根据x轴的位置获取当前所处位置tab的索引
 *
 * @param {Number} deltaX - 当前 x 轴的偏移量
 * @param {Number} indexWidth - 一个索引对应的宽度
 */
export const getIndexByDeltaX = (deltaX, indexWidth) => Math.floor(deltaX / indexWidth);

/**
 * @desc 根据x轴偏移量获取距离最近tab的索引
 *
 * @param {Number} deltaX - 当前 x 轴的偏移量
 * @param {Number} indexWidth - 一个索引对应的宽度
 * @param {Number} maxIndex - 最大的索引值
 */
export const getNearestIndexByDeltaX = (deltaX, indexWidth, maxIndex = 100) => {
  if (deltaX > 0) return 0;
  return inMaxMin(0, maxIndex, Math.round(Math.abs(deltaX) / indexWidth));
};

/**
 * @desc 获取索引在循环范围内的邻近的值
 *
 * @param {Number} idx - 新索引
 * @param {Number} min - 最小值
 * @param {Number} max - 最大值
 */
export const getSiblingIndex = (index, min, max) => {
  if (index === min && index === max) {
    return [index, index];
  } else if (index === min) {
    return [index, index + 1];
  } else if (index === max) {
    return [index - 1, index];
  }
  return [index - 1, index + 1];
};

/**
 *
 * @desc 获取居中对齐的tab索引
 *
 * @param {Number} idx - 当前tab的索引
 * @param {Number} oneScreenTabNum - 一屏可放下的tab数量
 * @param {Number} totalNum - tab总数量
 */
export const getCenteredScrollIndex = (idx, oneScreenTabNum, totalNum) => {
  const min = 0;
  const max = Math.max(0, totalNum - oneScreenTabNum);
  const median = Math.ceil(oneScreenTabNum / 2) - 1;
  const scrollToIdx = idx - median;
  const index = inMaxMin(min, max, scrollToIdx);
  return index;
};

/**
 * @desc 判断是否为合法的 Press 事件
 *
 * @param {Number} dx - x轴偏移量
 * @param {Number} dy - y轴偏移量
 */
export const isValidPress = (dx, dy) =>
  Math.abs(dx) <= PRESS_THRESHOLD && Math.abs(dy) <= PRESS_THRESHOLD;

/**
 * @desc 判断是否为合法的 Swipe 事件
 *
 * @param {Number} velocity - 加速度
 * @param {Number} direction - 移动距离
 */
export const isValidSwipe = (velocity, direction) => {
  return Math.abs(velocity) > VELOCITY_THRESHOLD && Math.abs(direction) < SWIPE_THRESHOLD;
};

/**
 * @desc 根据`tabLayout`累加算出每一个`tab`的`left`偏移量
 *
 * @param {Array} tabLayout - tab的布局属性，包含 x, y, width, height
 */
export const reduceTabLayoutLeft = tabLayout => {
  return tabLayout.reduce((acc, cur, index) => {
    if (index === 0) {
      return [{ ...cur, left: cur.x }];
    }
    const prev = acc[index - 1];
    const prevTotalX = prev.left + prev.width + prev.x;
    return [
      ...acc,
      {
        ...cur,
        left: prevTotalX + cur.x,
      },
    ];
  }, {});
};
