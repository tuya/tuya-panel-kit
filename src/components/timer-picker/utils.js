import { CoreUtils } from '../../utils';

const { toFixed } = CoreUtils;

export const getHourSelections = is12Hours => {
  const hours = [];
  for (let i = 0; i < 24; i++) {
    let label = '';
    if (is12Hours) {
      label = i === 0 ? 12 : i <= 12 ? i : i - 12;
    } else {
      label = toFixed(i, 2);
    }
    hours.push({
      value: i,
      label: label.toString(),
    });
  }
  return hours;
};

export const getMinsSelections = () => {
  const minutes = [];
  for (let i = 0; i < 60; i++) {
    minutes.push({
      value: i,
      label: toFixed(i, 2),
    });
  }
  return minutes;
};

export const getTimePrefixSelections = () => [
  {
    value: 'AM',
    label: 'AM',
  },
  {
    value: 'PM',
    label: 'PM',
  },
];

export const getPrefix = hour => (hour >= 12 ? 'PM' : 'AM');
