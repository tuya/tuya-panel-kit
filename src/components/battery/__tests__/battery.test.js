/**
 * @jest-environment jsdom
 */
import React from 'react';
import { mount } from 'enzyme';
import Battery from '../index'
import { ThemeUtils } from '../../../utils';

const { ThemeProvider } = ThemeUtils;

const setup = props => {
  const wrapper = mount(
    <ThemeProvider>
      <Battery {...props}/>
    </ThemeProvider>
    );
  return wrapper;
};

describe('Battery', () => {
  const origConsole = console.error;
  beforeEach(() => {
    console.error = () => {};
  });
  afterEach(() => {
    console.error = origConsole;
  });
  it('basic render', () => {
    const wrapper = setup({ onCalcColor: () => {return null},value: 10 });
    expect(wrapper.component).toMatchSnapshot();
  });
  it('render with color prop', () => {
    setup({ onCalcColor: () => {return {}} });
  });
  it('render with value == 50', () => {
    const wrapper = setup({ onCalcColor: () => {return null},value: 50 });
      expect(wrapper.component).toMatchSnapshot();
  });
  it('render with value == 0.5', () => {
    const wrapper = setup({ onCalcColor: () => {return null},value: 0.5 });
      expect(wrapper.component).toMatchSnapshot();
  });
});
