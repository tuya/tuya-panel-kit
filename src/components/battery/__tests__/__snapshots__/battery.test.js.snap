// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Battery basic render 1`] = `undefined`;

exports[`Battery render with value == 0.5 1`] = `undefined`;

exports[`Battery render with value == 50 1`] = `undefined`;
